File: topic-wise-problems/basic-array/two-sum/twoSum.go
// Optimised (hash map) solution
// Time Complexity: O(n)
// Space Complexity: O(n)

package twoSum

func twoSum(nums []int, target int) []int {
	numMap := make(map[int]int) // num : index

	for i, num := range nums {
		complement := target - num

		// Check if complement exists
		if idx, found := numMap[complement]; found {
			return []int{idx, i}
		}

		// Add current number to the map
		numMap[num] = i
	}

	return []int{}
}

File: topic-wise-problems/basic-array/two-sum/two_sum.py
# Optimised (hash map) solution
# Time Complexity: O(n)
# Space Complexity: O(n)

from typing import List

def twoSum(nums: List[int], target: int) -> List[int]:
    num_map = {}

    for i, num in enumerate(nums):
        complement = target - num

        # Check if complement exists
        if complement in num_map:
            return [num_map[complement], i]
        
        # Add current number to the map
        num_map[num] = i

    return []